File: Propuesta 1/SIMION/examples/pa2text/perl/pa2text.pl
#
# pa2text.cpp
# Reads the contents of a SIMION potential array file and writes the contents
# to standard output.
#
# This makes use of the SL Potential Array Tools library for
# the file I/O.
#
# David Manura, Scientific Instrument Services, Inc.
# $Revision$ $Date$ Created: 2003-11-14.
#

use strict;
use FindBin;
use lib "$FindBin::Bin/../../lib/perl";
use warnings;
use Getopt::Long;

use SIMION::PA;

my $headers = 1;
my $points  = 1;
my $verbose = 1;
my $output;

#-- read command-line parameters

my $result = GetOptions(
    "headers!" => \$headers,
    "points!" => \$points,
    # may be '-' for STDOUT
    "output=s" => \$output,
    "verbose!" => \$verbose
);

if(!$result || @ARGV < 1) {
    &print_usage();
}

my $infile = $ARGV[0];

my $outfile = $output || "$infile.txt";

unless($outfile eq '-') {
    # write to file (not standard output)
    print "Writing $outfile...\n";
    open(STDOUT, ">$outfile") or die "$! $outfile";
}

# read file
my $pa = new SIMION::PA(file => $infile);

# print header parameters
if($headers) {
    print "begin_header\n";
    print "mode=",        $pa->mode(),        "\n";
    print "symmetry=",    $pa->symmetry(),    "\n";
    print "max_voltage=", $pa->max_voltage(), "\n";
    print "nx=",          $pa->nx(),          "\n";
    print "ny=",          $pa->ny(),          "\n";
    print "nz=",          $pa->nz(),          "\n";
    print "mirror_x=",    $pa->mirror_x(),    "\n";
    print "mirror_y=",    $pa->mirror_y(),    "\n";
    print "mirror_z=",    $pa->mirror_z(),    "\n";
    print "field=",       $pa->field(),       "\n";
    print "ng=",          $pa->ng(),          "\n";
    print "end_header\n";
}

# print data points
if($points) {
    print "begin_points\n";
    # note: output in row-major order: points[nz][ny][nz]
    for my $z (0 .. $pa->nz()-1) {
        for my $y (0 .. $pa->ny()-1) {
            for my $x (0 .. $pa->nx()-1) {
                my ($elec, $pot) = $pa->point($x, $y, $z);
                $elec = $elec ? 1 : 0;
                if($verbose) {
                    print "x=$x, y=$y, z=$z, ";
                    print "elec=$elec, pot=$pot\n";
                }
                else { # terse format
                    print "$elec, $pot\n";
                }
            }
        }
    }
    print "end_points\n";
}


sub print_usage
{
    print "pa2text\n";
    print "  usage: perl pa2text.pl [options] <file>\n";
    print "where options can contain\n";
    print "  --[no]headers - whether to display headers\n";
    print "  --[no]points  - whether to display data points\n";
    print "  --output=filename - write to file.  If unspecified,\n";
    print "                      appends '.txt' to input file name.\n";
    print "                      Use '-' to write to standard output.\n";
    print "  --[no]verbose - verbose data point format\n";
    exit(1);
}

File: Propuesta 1/SIMION/examples/pa2text/python/pa2text.py
#
# pa2text.py
# Reads the contents of a SIMION potential array file and writes the contents
# to standard output.
#
# This makes use of the SL Potential Array Tools library for
# the file I/O.
#
# David Manura, Scientific Instrument Services, Inc.
# $Revision$ $Date$ Created 2003-11-14.
#

import sys
from SIMION.PA import *

def print_usage():
    print("usage: python pa2text.py <file>")
    sys.exit(1)

if len(sys.argv) < 2:
    print_usage()
    sys.exit(1)

path = sys.argv[1]

outfile = open(path + ".txt", "w")

#-- example reading
pa = PA(file = path)

# print header parameters
outfile.write("begin_header\n")
outfile.write("symmetry=" + pa.symmetry() + "\n")
outfile.write("max_voltage=" + str(pa.max_voltage()) + "\n")
outfile.write("nx=" + str(pa.nx()) + "\n")
outfile.write("ny=" + str(pa.ny()) + "\n")
outfile.write("nz=" + str(pa.nz()) + "\n")
outfile.write("mirror_x=" + str(pa.mirror_x()) + "\n")
outfile.write("mirror_y=" + str(pa.mirror_y()) + "\n")
outfile.write("mirror_z=" + str(pa.mirror_z()) + "\n")
outfile.write("field=" + pa.field_type() + "\n")
outfile.write("ng=" + str(pa.ng()) + "\n")
outfile.write("end_header\n")


outfile.write("begin_points\n")
for z in range(0, pa.nz()):
    for y in range(0, pa.ny()):
        for x in range(0, pa.nx()):
            (electrode, potential) = pa.point(x,y,z)
            electrode_str = electrode and "1" or "0"
            outfile.write(electrode_str + "," + str(potential) + "\n")
outfile.write("end_points\n")

outfile.close()



